src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/ConstantExpectedAnalyzer.UnmanagedHelper.cs

```diff
@@ -0,0 +1,350 @@
+                {
+                    if (Parameter.Type.SpecialType != subsetCandidate.Parameter.Type.SpecialType ||
+                        subsetCandidate is not UnmanagedConstantExpectedParameter subsetCandidateTParameter)
+                    {
+                        validationDiagnostics = Diagnostic.Create(CA1861.ConstantInvalidConstantRule, argument.Syntax.GetLocation(), Parameter.Type.ToDisplayString());
+                        return false;
+                    }
+
+                    if (!_helper.IsLessThan(subsetCandidateTParameter.Min, Min) && !_helper.IsLessThan(Max, subsetCandidateTParameter.Max))
+                    {
+                        //within range
+                        validationDiagnostics = null;
+                        return true;
+                    }
+                    validationDiagnostics = Diagnostic.Create(CA1861.ConstantOutOfBoundsRule, argument.Syntax.GetLocation(), Min.ToString(), Max.ToString());
+                    return false;
+                }
+
+                public override bool ValidateValue(IArgumentOperation argument, object? constant, [NotNullWhen(false)] out Diagnostic? validationDiagnostics)
+                {
+                    if (constant is not null && _helper.TryConvert(constant, out T value))
+                    {
+                        if (!_helper.IsLessThan(value, Min) && !_helper.IsLessThan(Max, value))
+                        {
+                            validationDiagnostics = null;
+                            return true;
+                        }
+                        validationDiagnostics = argument.CreateDiagnostic(CA1861.ConstantOutOfBoundsRule, Min.ToString(), Max.ToString());
+                        return false;
+                    }
+                    validationDiagnostics = argument.CreateDiagnostic(CA1861.ConstantInvalidConstantRule, Parameter.Type.ToDisplayString());
+                    return false;
+                }
+            }
+        }
+
+        private static bool TryConvertSignedInteger(object constant, out long integer)
+        {
+            try
+            {
+                if (constant is string or bool)
+                {
+                    integer = default;
+                    return false;
+                }
+                integer = Convert.ToInt64(constant);
+            }
+            catch
+            {
+                integer = default;
+                return false;
+            }
+            return true;
+        }
+        private static bool TryConvertUnsignedInteger(object constant, out ulong integer)
+        {
+            try
+            {
+                if (constant is string or bool)
+                {
+                    integer = default;
+                    return false;
+                }
+                integer = Convert.ToUInt64(constant);
```

Comment:
Is constant nullable?
If yes, then that is fine.
If no, I think ulong.Parse would be more appropriate.

Reply:
Instead of the Try-Catch I also recommend to use TryParse.

Reply:
constant is a primitive that is not a string and not a bool

Reply:
+1 for using TryParse, actually it seems to me simpler than keeping track of all this Min/Max values and checking if it is within the range, no?

Reply:
It is a boxed primitive, there is nothing to parse

Reply:
I meant not here with this design, instead of this approach could have used try parse from the beginning, as that would be complete change of the current design, I am not requesting this change. I was just wondering why did you choose this approach over `TryParse` directly

Reply:
The object value comes from AttributeData directly